odoo/tools/query.py

```diff
@@ -221,6 +221,18 @@ def __bool__(self):
         return bool(self.get_ids())
 
     def __len__(self):
+        if self._ids is None and not self.offset:
+            from_clause, where_clause, params = self.get_sql()
+            where_clause = where_clause or 'TRUE'
+            if self.limit:
+                # Special case to avoid counting every record in DB, which can
+                # be really slow. The result will be between 0 and limit.
+                query_str = f'SELECT FROM {from_clause} WHERE {where_clause} LIMIT {int(self.limit)}'
+                query_str = f"SELECT COUNT(*) FROM ({query_str}) t"
+            else:
+                query_str = f'SELECT COUNT(*) FROM {from_clause} WHERE {where_clause}'
+            self._cr.execute(query_str, params)
+            return self._cr.fetchone()[0]
```

Comment:
proposition to use [rowcount](https://www.psycopg.org/docs/cursor.html?highlight=count#cursor.rowcount) with an empty select. This leads to one less aggregate (for the count) on the execution plan than the same select count(*)

I checked the result of postgres and it doesn't send back the empty rows : 
`psql -d master -a -c "select from ir_ui_view;" -H` 
returns in html
```
<table border="1">
  <tr>
  </tr>
</table>
<p>(1911 rows)<br />
</p>
```


It is not standard sql but it is supported and documented in postgres.

```
query_str = f'SELECT FROM {from_clause} WHERE {where_clause}
if self.limit:
   query_str += f' limit {int(self.limit)}'
self._cr.execute(query_str, param)
return self._cr.rowcount
```

Reply:
I done a quick test this morning: it is much slower (more than 35 %) for the basic case (without limit).

Reply:
Probably because it is actually returning all empty rows.  That's what you get when you `cr.fetchall()`: a big list with empty tuples.  Possibly the HTML output is an optimization of the actual output...

Reply:
> 

ho damn... thanks for testing :)